### PR TITLE
[WIP] [JSX] Add newlines between elements, preserving spaces

### PR DESCRIPTION
**Pull request by rattrayalex**
Supersedes https://github.com/jlongster/prettier/pull/219 and intended to supersede https://github.com/jlongster/prettier/pull/220

Still hammering a few tricky trailing-whitespace situations out, but may be ready for 👀 nonetheless

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -2032,27 +2032,68 @@ function printJSXElement(path, options, print) {
...
             children.push(hardline);
+          } else if (/\s/.test(child.value)) {
+            children.push(" ");
+            children.push(softline);
           }
         } else {
           children.push(print(childPath));
         }
+
+        if (
```
Comment (rattrayalex) [inline src/printer.js]: @yamafaktory thoughts on this modified version?

File: src/printer.js [hunk 2]
```diff
@@ -2032,27 +2032,68 @@ function printJSXElement(path, options, print) {
...
+          if (!nextChild || nextChild.type === 'JSXElement') {
+            children.push(softline);
+          } else if (!nextChild.value.match(/\S/) && !nextChild.value.match(/\n/)) {
+            // children.push(softline);
+          }
+        }
+
+        if (typeof lastChild === "string" && lastChild.match(/\s$/)) {
+          if (children[lastIdx + 1].type === "line") {
+            children[lastIdx] = lastChild.replace(/\s$/, "{' '}");
```
Comment (rattrayalex) [inline src/printer.js]: this is an ugly hack... the most elegant solution I've been able to think of, though, would be to put something in the final printer that replaces any trailing whitespace within a JSX context with `{' '}`.

Comment (rattrayalex) [inline src/printer.js]: @jlongster would that approach be desirable/workable? thoughts?

File: src/printer.js (hunk 1 above)
Comment (yamafaktory) [inline src/printer.js]: @rattrayalex I like this approach which is covering a lot of use cases 👍

**Comment by jlongster** [comment]
I will look through the PR now and give feedback, but first I tried this out. It definitely works better. Thanks a lot!

Here are a few cases to talk about (I may have missed decisions about these earlier, but I don't think so).

```js
var x = <div>
    hello <strong>hi</strong> <foo>sdkflsdfjk</foo>
 </div>;
```

The above is turned into:

```js
var x = (
  <div>
    hello <strong>hi</strong>
    <foo>sdkflsdfjk</foo>
  </div>
);
```

This example never had to break -- should `foo` be moved down to the next line? It seems arbitrary that `strong` wasn't but `foo` was. What is the exact rule for moving things down?

Well, the above the case illustrates my question. I thought I had other cases but they're all pretty much the same as above.

**Comment by jlongster** [comment]
I wonder if we should create a document that described various rules we use for breaking things. I wouldn't want this to be very long, but some short succinct descriptions about how we deal with popular patterns might give a good overview and be a reference we can point to when people open new issues. @rattrayalex If you want to, can you start such a document in `docs`? Don't worry if you don't have time / don't want to. I can start it soon myself. Seeing the JSX algorithm described in words would be nice though.

**Comment by jlongster** [comment]
Actually, don't worry about the document. We're changing things pretty quickly right now and it'll get out-of-date fast so it's not worth it. It would be good at least to explain in #202 any edge cases you find and flesh it out a little so we have it as a reference.

File: src/printer.js [hunk 3]
```diff
@@ -1074,7 +1074,7 @@ function genericPrintNoParens(path, options, print) {
               n.selfClosing ? line : softline
             ])
           ),
-          n.selfClosing ? "/>" : ">"
+          n.selfClosing ? "/>" : ">",
```
Comment (jlongster) [inline src/printer.js]: Currently we don't use trailing commas. Although this will be stripped out when we reprint the file so doesn't really matter :)

File: src/printer.js (hunk 2 above)
Comment (rattrayalex) [inline src/printer.js]: Okay, I think what I want to try is adding an additional node type, `'jsx-whitespace'`. It's unfortunate to have a whole node type just for this edge case, but then again it's essentially a whitespace-syntax-tree, and, well, JSX has its own whitespace semantics – so I think it's a legitimate use-case.

File: src/printer.js (hunk 3 above)
Comment (rattrayalex) [inline src/printer.js]: Hah, I should probably get in the habit of actually *using* prettier while I'm working on it...

Comment (jlongster) [inline src/printer.js]: I actually don't either yet. I started to a while back but it was too unstable, it's probably fine now. I'm fine with waiting a little longer until we enforce it (need to figure out tooling first anyway)

**Comment by jlongster** [comment]
I'm having a hard time understanding this code, mainly because I haven't used JSX very much. It seems tricky, so we need to have comments documenting why each case is handled the way it is. It would also help a lot with the review (I think that's what I was looking for when I was talking about docs above). Can you add thorough comments in the JSX printing code? I think I'd have an easier time understanding why these special cases exist, as would future maintainers.

File: src/printer.js (hunk 2 above)
Comment (jlongster) [inline src/printer.js]: Sorry, missed this comment. I thought JSX removes whitespace at the beginning and end of a line? What is this whitespace significant? Can you paste some JSX code to help me get the context of this?

I'd prefer to avoid the printer being aware of JSX, but once I understand why this is needed we can figure something out.

**Comment by rattrayalex** [comment]
Haha yeah some comments would indeed be helpful 😄 I'll do that.

I think there may be something "wrong" though... 

What's happening is that there's only one `group` per element, including its opening element, its children, and its closing element. That group is a `multilineGroup`, so if anything is split, *all* children are split into their own lines, if there are `line`s on which to split. So currently, most element-splitting logic here is done by inserting `softline`s selectively. 

Mostly to specifically support the sort of use-case you highlighted, those `softline`s are only inserted directly between node elements, not between node elements and text. (They're also inserted after a space-only text element, which I believe only occurs between node elements).

It would probably be preferable to add `softline`s between every kind of element, node and text, and have better grouping logic. Everything I've tried to that effect has come out pretty ugly, but my group-foo may be too basic at this point. 

The challenge is that this:
```jsx
<div
  long="attrs"
><child /></div>
```
should always split the child out with an indent, which means a single `multilineGroup` around the whole element (what we have now). But if you wrap the children code in its own (multiline)group, that breaks. 

Any ideas?

File: src/printer.js (hunk 2 above)
Comment (rattrayalex) [inline src/printer.js]: Yeah, sorry, it's fairly confusing – see https://github.com/jlongster/prettier/issues/201

File: src/printer.js [hunk 4]
```diff
@@ -2032,27 +2032,68 @@ function printJSXElement(path, options, print) {
...
+            } else if (endSpace) {
+              children.push(" ");
+            }
 
-            children.push(
-              beginBreak ? hardline : "",
-              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
-              endBreak ? hardline : ""
-            );
           } else if (/\n/.test(child.value)) {
```
Comment (jlongster) [inline src/printer.js]: I'm also confused about something here: how does this work? JSX elements are not split up per-line. For example:

```js
<div>
  foo
  bar
  baz
</div>
```

The text contents are represented by this node:

```js
Node {
  type: 'JSXText',
  start: 5,
  end: 24,
  loc:
  SourceLocation {
    start: Position { line: 1, column: 5 },
    end: Position { line: 5, column: 0 } },
  extra: null,
  value: '\n  foo\n  bar\n  baz\n'
}
```

I don't understand when we are able to just test for a newline and emit only a line. Shouldn't we handle newlines more elegantly?

Note that we cannot just output this text. For the printer to accurately measure, we need to convert them into line objects:

```
join(line, node.value.split("\n"))
```

Something like that. I swear we used to do that but I don't see it anywhere in here.

File: src/printer.js (hunk 2 above)
Comment (rattrayalex) [inline src/printer.js]: tl;dr JSX parses these two cases differently, but `prettier` doesn't:

```jsx
<div /> <div />
```
```jsx
<div />
 <div />
```
one renders a space between the divs, the other does not.

**Comment by jlongster** [comment]
> It would probably be preferable to add softlines between every kind of element, node and text, and have better grouping logic. Everything I've tried to that effect has come out pretty ugly, but my group-foo may be too basic at this point.

Aah I see. Can this be summarized as: we should always try to break the children of a JSX element first before breaking the attributes?

File: src/printer.js (hunk 2 above)
Comment (rattrayalex) [inline src/printer.js]: For illustrative purposes (or in case you're deeply curious about this) [JSX parses whitespace with newlines completely differently from normal JS](https://github.com/babel/babylon/blob/master/src/plugins/jsx/index.js#L83).

**Comment by jlongster** [comment]
> What's happening is that there's only one group per element, including its opening element, its children, and its closing element.

Also, that does seem a little weird and we probably need to change it so it adds a group per JSX element so deeper elements can be broken without forcing all parents to break.

**Comment by rattrayalex** [comment]
Ahhh yes, I was starting to come to the same conclusion as I typed out that explanation. The "when you're not doing outer-to-inner, you might need conditionalGroup" sounds like a nice heuristic. I've used it before, but also abused it, as you may recall 😉

I'll give that a try!

File: src/printer.js (hunk 4 above)
Comment (rattrayalex) [inline src/printer.js]: Hah, that's funny – I actually *just* linked to babylon's jsx-newline parsing code in the other thread, which explains how this happens. 

I'll make sure test cases include examples like that btw.

the case you directly linked to occurs when a jsxtextnode contains a newline, but no non-space characters – basically, when you have at least one newline between JSX elements.

Comment (rattrayalex) [inline src/printer.js]: ... which makes me realize, we need to add a `literalline` for each `\n` found, not just a single `hardline`, I think... will add tests for that.

Comment (rattrayalex) [inline src/printer.js]: ... or should `prettier` enforce only a single, or up to two, newlines between jsx elements?

**Comment by jlongster** [comment]
Cool! Yeah, it's a "use only in really special cases" type thing but this is a special case. Just make sure you share documents as much as possible across the groups, i.e.

```js
const childrenPrinted = path.call(print, "children");
return conditionalGroup([
  concat([" ", childrenPrinted, " "]),
  concat([hardline, childrenPrinted, hardline]),
])
```

instead of 

```js
return conditionalGroup([
  concat([" ", path.call(print, "children"), " "]),
  concat([hardline, path.call(print, "children"), hardline]),
])
```

Because the latter we print the expression twice and duplicate the children document, which is really bad both for CPU and memory.

File: src/printer.js (hunk 4 above)
Comment (rattrayalex) [inline src/printer.js]: answering my own question, up to two sounds good.

**Comment by jlongster** [comment]
@rattrayalex This is getting into parts of the API that we probably need to polish some more, but one more thing. You can forcefully break a document wrapping it in a group and passing an option: `group(doc, { shouldBreak: true })`. This will force everything in `doc` to break. (If it's not clear: a document is any piece of the "intermediate representation", i.e. anything returned from `print`).

This is useful with `conditionalGroup` because you can call `print` on an arbitrary expression but forcefully break it. That's how you force a child expression to break before the parent.

**Comment by rattrayalex** [comment]
That's very helpful as well! Cool I'll get cracking on that.

**Comment by jlongster** [comment]
If you need any real-time help, feel free to jump on gitter! You're doing some of the most complicated work so I'm happy to work through it with you. You can also check out `printArgumentsList` which is another place we use this technique (which allows the last arg of a function call to expand across lines before breaking all params, which is the same break-child-first problem).

**Comment by rattrayalex** [comment]
Very close, sharing progress. Found a weird edge case.

**Comment by yamafaktory** [comment]
@rattrayalex & @jlongster really interesting discussion and progress, looking forward to seeing how it will go 👍

File: tests/prettier/__snapshots__/jsfmt.spec.js.snap [hunk 5]
```diff
@@ -80,13 +82,127 @@ const comp5 = <div>Keep it on one line.</div>;
...
+    {\' \'}foo
+  </span>
+);
+
+before_break2 = (
+  <span>
+    <span
+      barbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbar
+    />
+    {\' \'}foo
```
Comment (rattrayalex) [inline tests/prettier/__snapshots__/jsfmt.spec.js.snap]: `{\' \'}` could also be split on its own line if desired

**Comment by rattrayalex** [comment]
Okay, having slept on this a bit, I think we should step back and ask what formatting we *want* for JSX. 

Most of my uncertainty revolves around siblings on the same line, and as a special case, mixed text and node siblings. Eg; 

(1) 
```jsx
<div>
  <Icon type="foo" /><Icon type="bar" />
</div>
```
(2)
```jsx
<div>
  <Icon type="foo" /><Icon type="bar" />
  <Icon type="foo" /><Icon type="bar" />
  <Icon type="foo" /><Icon type="bar" />
</div>
```
(3)
```jsx
<div>
  <Icon type="foo" /> This is a paragraph with <strong>text</strong>.
  <div with="long attrs, let's pretend" />
</div>
```
(4)
```jsx
<div>
  A bunch of text here...
  <br /><hr /><br />
  And a bunch more <strong>text</strong> here...
</div>
```

In all the cases above, everything within the div is a sibling. So a typical group approach would break onto its own line. Example (4) is particularly, telling, which would be broken out into:

```jsx
<div>
  A bunch of text here...
  <br />
  <hr />
  <br />
  And a bunch more{' '}
  <strong>text</strong>
  {' '}here...
</div>
```

That seems clearly wrong in the sentence-with-strong examples. But what about the examples without text, like (2)? My bunch is that in (2), all Icon's should be on their own line, and that if the same happened to (1), that would be acceptable – but is it?

If (2) should remain printed as-written, then we probably want to wrap each actual line in its own `group` (or maybe `multilineGroup`). That sounds somewhat tricky and hacky but probably doable.

If (2) should break but (3) should not, then we probably want to treat text with jsx siblings as belonging to an invisible `<span>` parent, which also sounds tricky (wrap text-and-jsx-siblings in a group, probably only per-line). 

Thoughts? Should we ask **gaearon** for input on this?

**Comment by rattrayalex** [comment]
Okay, I've gone and implemented the above algorithm (wrap each line in a `group`). This might sound like it defeats the purpose of `prettier`, but I think it makes sense in the context of JSX because you can essentially have multiple "statements" per line, which is impossible in JS (without semicolons). 

It resulted in generally simpler code, though there still is a lot going on. 

There's one lingering `{' '}` related bug, but that should be fixable in an isolated way. 

Otherwise, I think this is ready for a careful review.

File: tests/prettier/__snapshots__/jsfmt.spec.js.snap [hunk 6]
```diff
@@ -129,13 +260,115 @@ long_open = (
...
+
+leave_opening = (
+  <BaseForm
+    url=\"/auth/google\"
+    method=\"GET\"
+    colour=\"blue\"
+    size=\"large\"
+    submitLabel=\"Sign in with Google\"
+  >
+
```
Comment (rattrayalex) [inline tests/prettier/__snapshots__/jsfmt.spec.js.snap]: This is a bug; should be `{' '}`. Needs to be fixed before shipping.

**Comment by rattrayalex** [comment]
@vjeux not sure you were tagged on this, would love your 👀 if you have the time and inclination!

**Comment by vjeux** [comment]
I don't know what you did, but it's actually working like magic on the Nuclide codebase. Most of the differences I see are actually better with your algorithm!

Two things that are probably best addressed in follow ups.

- [ ] We shouldn't output parenthesis inside of function calls.

image

- [ ] There's a bunch of cases where inside of a JSXExpression, the output is very whitespace and linebreaks heavy. We should be more aggressive at condensing it

image

**Comment by rattrayalex** [comment]
Awesome to hear!!

Could you link to the source of those examples so I can add them as tests?
They may both have a simple quick fix that I'd happily roll into this PR.

**Comment by rattrayalex** [comment]
K, I added non-paren cases for `CallExpression` and `ConditionalExpression`, with tests. Ternaries still need some fixing, but that should be in a follow-up.

**Comment by vjeux** [comment]
Oh sorry I missed your question. Glad you figured it out!

Can't wait to get this landed <3

**Comment by existentialism** [comment]
Nice work @rattrayalex!

File: src/printer.js [hunk 7]
```diff
@@ -2019,71 +2022,170 @@ function printMemberChain(node, options, print) {
...
+
+  return children;
+}
+
 function printJSXElement(path, options, print) {
   const n = path.getValue();
-  var openingLines = path.call(print, "openingElement");
 
-  let elem;
+  // turn <div></div> into <div />
```
Comment (jlongster) [inline src/printer.js]: Are we sure we want to do this? It could be annoying if someone is in the middle of writing code and saves the file which autoformats the file. To add content they need to break it up again.

Comment (rattrayalex) [inline src/printer.js]: Hmm, that does sound annoying... though I imagine there may be many such situations with a formatter like `prettier`, no?

It's definitely desired for a final-state program, though. `></div>` is pretty universally considered bad JSX style as far as I'm aware. 

In case it's any consolation, it won't break up if there's a space, so if you're writing along and have `<div> </div>`, that will stay as-is (because the space is significant).

File: src/printer.js [hunk 8]
```diff
@@ -2019,71 +2022,170 @@ function printMemberChain(node, options, print) {
...
+    childrenGroupedByLine.push(children[0]);
 
-    const firstChild = children[(0)];
-    const lastChild = util.getLast(children);
-    const beginBreak = firstChild && firstChild.type === "line";
-    const endBreak = lastChild && lastChild.type === "line";
+    if (!hasAnyHardLine && hasHardLine(children[0])) hasAnyHardLine = true;
+  } else {
+    // group by line, recording if there was a hardline.
+    childrenGroupedByLine.push(group(concat([])));
```
Comment (jlongster) [inline src/printer.js]: Why is this needed?

Comment (rattrayalex) [inline src/printer.js]: this initializes the first group... it's kind of ugly code, and used to be prettier until I handled the jsx-whitespace case above with this `if`

File: src/printer.js (hunk 7 above)
Comment (jlongster) [inline src/printer.js]: Ok. I think it's good. I agree that there are places we consolidate things and it's a balance of being helpful and not annoying. I think this is is a good idea.

Comment (rattrayalex) [inline src/printer.js]: Cool. I'll be thinking about this though, because it's almost guaranteed to annoy at least some of the users some of the time... would be great to have a simple fix...

Maybe in the documentation, mentioning something like "If you have format-on-save, and you're adding a bunch of empty JSX elements to fill in later, you might want to add an empty `{}` within each one" could be helpful. Though, of course, the need to do that is certainly a smell...

File: src/printer.js (hunk 8 above)
Comment (rattrayalex) [inline src/printer.js]: I think I'll do this another way...

File: src/printer.js [hunk 9]
```diff
@@ -2048,71 +2050,198 @@ function printMemberChain(node, options, print) {
...
+        children.push(softline);
+      }
+    }
+  }, "children");
+
+  return children;
+}
+
+// JSX expands children from the inside-out, instead of the outside-in.
+// This is both to break children before attributes,
```
Comment (rattrayalex) [inline src/printer.js]: @jlongster look good?

File: src/printer.js [hunk 10]
```diff
@@ -2048,71 +2050,198 @@ function printMemberChain(node, options, print) {
...
+  if (children.length === 1) {
+    if (!hasAnyHardLine && hasHardLine(children[0])) hasAnyHardLine = true;
 
-    var mostChildren = children.slice(0, -1);
-    var closingLines = path.call(print, "closingElement");
+    // no need for groups, and a lone jsx whitespace doesn't work otherwise
+    childrenGroupedByLine = [ hardline, children[0] ];
+  } else {
+    //  prefill leading newline, and initialize the first line's group
+    childrenGroupedByLine = [ hardline, group(concat([])) ];
```
Comment (rattrayalex) [inline src/printer.js]: @jlongster you'd asked about this before... still one of the uglier bits in this PR IMO, but the only prettier ways I can think of would sacrifice perf. look ok to you?

Comment (jlongster) [inline src/printer.js]: Yep! Totally fine and I'm fine sacrificing some readability for perf.

File: src/printer.js (hunk 9 above)
Comment (jlongster) [inline src/printer.js]: yep this looks awesome! great work!

File: src/printer.js (hunk 7 above)
Comment (jlongster) [inline src/printer.js]: It's up to us to decide what benefits most users, so even if we have to explain something like that, I think the feature can be beneficial to more users than not. So I think it's ok. There are lots of places where we've had to make decisions like this and if enough people complain we can tweak things.

**Comment by rattrayalex** [comment]
Feel free to merge when ready, @jlongster !

**Comment by jlongster** [comment]
Truly amazing work @rattrayalex, I'm super impressed!

**Comment by tychota** [comment]
Wow!
